File: apps/api/app/api/failures.py
"""Failure investigation and approval routes — Slice 6.

Both routers live here because they are two halves of one slice and share
their error translation. They are mounted at separate prefixes.

**Permissions, checked against migration 002 before writing:**

    failure.view              Chemist, Engineer, Lead, Director, QA
    failure.create            Chemist, Engineer
    failure.investigate       Chemist, Engineer     hypotheses and evidence
    failure.accept_root_cause **Lead alone**
    failure.close             **Lead alone**

🔴 `failure.accept_root_cause` IS THE WHOLE POINT OF THE SPLIT.

`CLAUDE.md` §7: an AI hypothesis is not an accepted root cause, and only a
human moves it. The Chemist and Engineer who investigate hold
`failure.investigate` and NOT `failure.accept_root_cause` — so the people
generating hypotheses are structurally unable to conclude one, and MSD
(Slice 7) inherits that boundary by having no permissions at all.

The administrator holds neither, deliberately: migration 002 says
"administering the system is not the same authority as making a technical
decision".
"""

from __future__ import annotations

import datetime as dt
import uuid
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.core.security import PermissionDenied, Principal, get_db, require_permission
from app.core.tenancy import CrossTenantReferenceError
from app.domains.approvals.service import (
    ApprovalError,
    ApprovalNotFoundError,
    ApprovalStateError,
    IncompatibleDutyError,
    decide_step,
    get_route,
    open_route,
    pending_steps_for,
    route_for_entity,
)
from app.domains.approvals.service import DecisionInput as ApprovalDecisionInput
from app.domains.failures.service import (
    ActionInput,
    DriverInput,
    EvidenceInput,
    FailureError,
    FailureInput,
    FailureNotFoundError,
    FailureStateError,
    HypothesisInput,
    accept_root_cause,
    close_failure,
    get_failure,
    link_evidence,
    list_failures,
    open_failure,
    raise_action,
    record_driver,
    record_evidence,
    record_hypothesis,
    reject_hypothesis,
)

router = APIRouter()
approvals_router = APIRouter()

__all__ = ["approvals_router", "router"]


class FailureCreate(BaseModel):
    project_id: uuid.UUID
    failure_code: str = Field(min_length=3, max_length=50)
    title: str = Field(min_length=3, max_length=200)
    description: str | None = None
    severity: str = Field(default="major", pattern="^(critical|major|minor)$")
    test_id: uuid.UUID | None = None
    formula_version_id: uuid.UUID | None = None
    batch_id: uuid.UUID | None = None


class HypothesisCreate(BaseModel):
    possible_cause: str = Field(min_length=3, max_length=2000)
    mechanism: str | None = Field(default=None, max_length=2000)
    confidence: str = Field(default="medium", pattern="^(low|medium|high)$")
    source: str | None = Field(default=None, max_length=500)
    # An MSD-proposed hypothesis must SAY it is one. §7 rests on the
    # distinction, and a default of `human` means the label can only be
    # wrong by a caller asserting it.
    origin: str = Field(default="human", pattern="^(human|msd)$")


class EvidenceCreate(BaseModel):
    evidence_type: str = Field(
        pattern="^(previous_experiment|literature|batch_deviation|material_lot_issue|"
        "test_trend|photograph|other)$"
    )
    summary: str = Field(min_length=3, max_length=2000)
    detail: str | None = None
    referenced_entity_type: str | None = Field(
        default=None, pattern="^(test|batch|material_lot|formula_version|document)$"
    )
    referenced_entity_id: uuid.UUID | None = None
    source_reference: str | None = Field(default=None, max_length=500)
    origin: str = Field(default="human", pattern="^(human|msd)$")


class EvidenceLink(BaseModel):
    evidence_id: uuid.UUID
    relationship: str = Field(default="supports", pattern="^(supports|contradicts|inconclusive)$")
    note: str | None = Field(default=None, max_length=1000)


class RootCauseAcceptance(BaseModel):
    hypothesis_id: uuid.UUID
    # Required. Accepting a root cause is a technical decision and an
    # unexplained one cannot be reviewed later.
    rationale: str = Field(min_length=3, max_length=2000)


class HypothesisRejection(BaseModel):
    reason: str = Field(min_length=3, max_length=2000)


class ActionCreate(BaseModel):
    action_type: str = Field(
        pattern="^(formula_revision|repeat_test|process_change|material_change|start_doe|other)$"
    )
    description: str = Field(min_length=3, max_length=2000)
    assigned_to: uuid.UUID | None = None
    due_date: dt.date | None = None


class ClosureCreate(BaseModel):
    summary: str = Field(min_length=3, max_length=2000)


class DriverCreate(BaseModel):
    driver_type: str = Field(
        pattern="^(failure|requirement|optimization|cost|regulatory|customer_request|other)$"
    )
    reason: str = Field(min_length=3, max_length=2000)
    failure_id: uuid.UUID | None = None
    requirement_id: uuid.UUID | None = None


class RouteOpen(BaseModel):
    project_id: uuid.UUID
    entity_type: str = Field(
        pattern="^(test|formula_version|validation|pilot|qualification|product_release)$"
    )
    entity_id: uuid.UUID
    authority_level: str = Field(
        pattern="^(preliminary|development|controlled|validation|qualification|release)$"
    )


class StepDecision(BaseModel):
    decision: str = Field(
        pattern="^(approve|approve_with_condition|return_for_correction|"
        "request_retest|reject|escalate|request_additional_test)$"
    )
    condition_text: str | None = Field(default=None, max_length=2000)
    rationale: str | None = Field(default=None, max_length=2000)


def _missing(exc: Exception) -> HTTPException:
    return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))


def _conflict(exc: Exception) -> HTTPException:
    return HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


def _invalid(exc: Exception) -> HTTPException:
    return HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc))


# ---------------------------------------------------------------------------
# Failures
# ---------------------------------------------------------------------------


@router.get("", tags=["failures"])
def get_failures(
    project_id: uuid.UUID | None = Query(default=None),
    status_filter: str | None = Query(default=None, alias="status"),
    principal: Principal = Depends(require_permission("failure.view")),
    session: Session = Depends(get_db),
) -> list[dict[str, Any]]:
    """The investigation queue.

    `open_actions` and `has_root_cause` come back on every row because
    they are what makes it actionable — §11 requires counts to represent
    items needing action, not total rows.
    """
    return list_failures(
        session,
        organization_id=principal.organization_id,
        project_id=project_id,
        status=status_filter,
    )


@router.post("", status_code=status.HTTP_201_CREATED, tags=["failures"])
def post_failure(
    payload: FailureCreate,
    principal: Principal = Depends(require_permission("failure.create")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    try:
        return open_failure(
            session,
            project_id=payload.project_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            spec=FailureInput(
                failure_code=payload.failure_code,
                title=payload.title,
                description=payload.description,
                severity=payload.severity,
                test_id=payload.test_id,
                formula_version_id=payload.formula_version_id,
                batch_id=payload.batch_id,
            ),
        )
    except FailureNotFoundError as exc:
        raise _missing(exc) from exc
    except (FailureError, CrossTenantReferenceError) as exc:
        raise _invalid(exc) from exc


@router.get("/{failure_id}", tags=["failures"])
def get_one_failure(
    failure_id: uuid.UUID,
    principal: Principal = Depends(require_permission("failure.view")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """One investigation: hypotheses with their evidence, and how each
    piece bears on each hypothesis.

    `accepted_root_cause` is returned separately, and is `null` until a
    Lead accepts one. A screen must be able to show "no root cause yet"
    as distinct from "here are some ideas".
    """
    try:
        return get_failure(
            session, failure_id=failure_id, organization_id=principal.organization_id
        )
    except FailureNotFoundError as exc:
        raise _missing(exc) from exc


@router.post("/{failure_id}/hypotheses", status_code=status.HTTP_201_CREATED, tags=["failures"])
def post_hypothesis(
    failure_id: uuid.UUID,
    payload: HypothesisCreate,
    principal: Principal = Depends(require_permission("failure.investigate")),
    session: Session = Depends(get_db),
) -> dict[str, str]:
    """Propose a possible cause. Always `proposed` — never accepted.

    `failure.investigate` deliberately does NOT carry
    `failure.accept_root_cause`: whoever generates hypotheses must not be
    able to conclude one.
    """
    try:
        hypothesis_id = record_hypothesis(
            session,
            failure_id=failure_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            spec=HypothesisInput(**payload.model_dump()),
        )
    except FailureNotFoundError as exc:
        raise _missing(exc) from exc
    except FailureStateError as exc:
        raise _conflict(exc) from exc
    return {"id": str(hypothesis_id)}


@router.post("/{failure_id}/evidence", status_code=status.HTTP_201_CREATED, tags=["failures"])
def post_evidence(
    failure_id: uuid.UUID,
    payload: EvidenceCreate,
    principal: Principal = Depends(require_permission("failure.investigate")),
    session: Session = Depends(get_db),
) -> dict[str, str]:
    try:
        evidence_id = record_evidence(
            session,
            failure_id=failure_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            spec=EvidenceInput(**payload.model_dump()),
        )
    except FailureNotFoundError as exc:
        raise _missing(exc) from exc
    return {"id": str(evidence_id)}


@router.post(
    "/{failure_id}/hypotheses/{hypothesis_id}/evidence",
    status_code=status.HTTP_201_CREATED,
    tags=["failures"],
)
def post_evidence_link(
    failure_id: uuid.UUID,
    hypothesis_id: uuid.UUID,
    payload: EvidenceLink,
    principal: Principal = Depends(require_permission("failure.investigate")),
    session: Session = Depends(get_db),
) -> dict[str, str]:
    """Attach an observation to a hypothesis, saying HOW it bears on it.

    `contradicts` is an ordinary value here. An investigation that could
    only record confirming evidence cannot rule anything out.
    """
    try:
        link_id = link_evidence(
            session,
            hypothesis_id=hypothesis_id,
            evidence_id=payload.evidence_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            relationship=payload.relationship,
            note=payload.note,
        )
    except FailureNotFoundError as exc:
        raise _missing(exc) from exc
    except FailureError as exc:
        raise _invalid(exc) from exc
    return {"id": str(link_id)}


@router.post("/{failure_id}/root-cause", tags=["failures"])
def post_root_cause(
    failure_id: uuid.UUID,
    payload: RootCauseAcceptance,
    principal: Principal = Depends(require_permission("failure.accept_root_cause")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """🔴 Promote a hypothesis to THE accepted root cause.

    The Lead alone. §7: only a human moves a hypothesis to `accepted`,
    and the people who propose them cannot.
    """
    try:
        return accept_root_cause(
            session,
            failure_id=failure_id,
            hypothesis_id=payload.hypothesis_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            rationale=payload.rationale,
        )
    except FailureNotFoundError as exc:
        raise _missing(exc) from exc
    except FailureStateError as exc:
        raise _conflict(exc) from exc
    except FailureError as exc:
        raise _invalid(exc) from exc


@router.post("/{failure_id}/hypotheses/{hypothesis_id}/rejection", tags=["failures"])
def post_hypothesis_rejection(
    failure_id: uuid.UUID,
    hypothesis_id: uuid.UUID,
    payload: HypothesisRejection,
    principal: Principal = Depends(require_permission("failure.investigate")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Rule a hypothesis out, with the reason the next investigator needs."""
    try:
        return reject_hypothesis(
            session,
            hypothesis_id=hypothesis_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            reason=payload.reason,
        )
    except FailureStateError as exc:
        raise _conflict(exc) from exc
    except FailureError as exc:
        raise _invalid(exc) from exc


@router.post("/{failure_id}/actions", status_code=status.HTTP_201_CREATED, tags=["failures"])
def post_action(
    failure_id: uuid.UUID,
    payload: ActionCreate,
    principal: Principal = Depends(require_permission("failure.investigate")),
    session: Session = Depends(get_db),
) -> dict[str, str]:
    try:
        action_id = raise_action(
            session,
            failure_id=failure_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            spec=ActionInput(**payload.model_dump()),
        )
    except FailureNotFoundError as exc:
        raise _missing(exc) from exc
    except CrossTenantReferenceError as exc:
        raise _invalid(exc) from exc
    return {"id": str(action_id)}


@router.post("/{failure_id}/closure", tags=["failures"])
def post_closure(
    failure_id: uuid.UUID,
    payload: ClosureCreate,
    principal: Principal = Depends(require_permission("failure.close")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Close the investigation, with its conclusion.

    Refused while corrective actions are outstanding: closing would leave
    them where no queue will surface them again.
    """
    try:
        return close_failure(
            session,
            failure_id=failure_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            summary=payload.summary,
        )
    except FailureNotFoundError as exc:
        raise _missing(exc) from exc
    except FailureStateError as exc:
        raise _conflict(exc) from exc
    except FailureError as exc:
        raise _invalid(exc) from exc


# ---------------------------------------------------------------------------
# Approvals — the shared engine
# ---------------------------------------------------------------------------


@approvals_router.get("/queue", tags=["approvals"])
def get_approval_queue(
    principal: Principal = Depends(require_permission("test.view")),
    session: Session = Depends(get_db),
) -> list[dict[str, Any]]:
    """Steps this caller can decide RIGHT NOW.

    Not every undecided step — only those whose turn has come. A queue
    listing work that is still blocked by an earlier group would show
    items the holder cannot act on, and §11 requires a count to mean
    items needing action BY THE HOLDER.

    The caller's own permission set does the filtering, so this endpoint
    needs no permission of its own beyond being able to see tests at all.
    """
    return pending_steps_for(
        session,
        organization_id=principal.organization_id,
        held_permissions=principal.permissions,
    )


@approvals_router.post("", status_code=status.HTTP_201_CREATED, tags=["approvals"])
def post_route(
    payload: RouteOpen,
    principal: Principal = Depends(require_permission("test.plan", "formula.submit")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Open an approval route by snapshotting the template for an authority."""
    try:
        return open_route(
            session,
            organization_id=principal.organization_id,
            project_id=payload.project_id,
            entity_type=payload.entity_type,
            entity_id=payload.entity_id,
            authority_level=payload.authority_level,
            actor_id=principal.user_id,
        )
    except ApprovalNotFoundError as exc:
        raise _missing(exc) from exc
    except ApprovalStateError as exc:
        raise _conflict(exc) from exc
    except ApprovalError as exc:
        raise _invalid(exc) from exc


@approvals_router.get("/{route_id}", tags=["approvals"])
def get_one_route(
    route_id: uuid.UUID,
    principal: Principal = Depends(require_permission("test.view")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """One route, its snapshotted steps, and who is being waited on.

    `awaiting` is what rule 12 of the traffic light needs in order to
    name the actual approver, rather than saying only "awaiting the next
    approver".
    """
    try:
        return get_route(session, route_id=route_id, organization_id=principal.organization_id)
    except ApprovalNotFoundError as exc:
        raise _missing(exc) from exc


@approvals_router.get("/entity/{entity_type}/{entity_id}", tags=["approvals"])
def get_route_for_entity(
    entity_type: str,
    entity_id: uuid.UUID,
    principal: Principal = Depends(require_permission("test.view")),
    session: Session = Depends(get_db),
) -> dict[str, Any] | None:
    """The open route for a record, or null.

    Null is an ordinary answer, not a 404: a test still in execution has
    nothing to approve, and that is not a missing resource.
    """
    return route_for_entity(
        session,
        organization_id=principal.organization_id,
        entity_type=entity_type,
        entity_id=entity_id,
    )


@approvals_router.post("/{route_id}/steps/{step_id}/decision", tags=["approvals"])
def post_step_decision(
    route_id: uuid.UUID,
    step_id: uuid.UUID,
    payload: StepDecision,
    principal: Principal = Depends(require_permission("test.view")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Decide one step.

    🔴 THE PERMISSION IS THE STEP'S, NOT THIS ENDPOINT'S.

    `require_permission("test.view")` here is only a floor — the engine
    checks the permission the SNAPSHOTTED STEP names, which is the
    template's decision and not the router's. Hard-coding a permission
    here would mean the route's own definition of who may sign was
    advisory, and a template could never introduce a step this endpoint
    had not anticipated.

    403 covers two different refusals and says which: the caller lacks
    the step's permission, or holds it and is disqualified by their own
    earlier involvement (ADR-019).
    """
    try:
        return decide_step(
            session,
            route_id=route_id,
            step_id=step_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            held_permissions=principal.permissions,
            spec=ApprovalDecisionInput(
                decision=payload.decision,
                condition_text=payload.condition_text,
                rationale=payload.rationale,
            ),
        )
    except ApprovalNotFoundError as exc:
        raise _missing(exc) from exc
    except IncompatibleDutyError as exc:
        raise PermissionDenied(str(exc)) from exc
    except ApprovalStateError as exc:
        raise _conflict(exc) from exc
    except ApprovalError as exc:
        raise _invalid(exc) from exc


@approvals_router.post(
    "/versions/{formula_version_id}/drivers",
    status_code=status.HTTP_201_CREATED,
    tags=["approvals"],
)
def post_version_driver(
    formula_version_id: uuid.UUID,
    payload: DriverCreate,
    principal: Principal = Depends(require_permission("formula.clone", "formula.create")),
    session: Session = Depends(get_db),
) -> dict[str, str]:
    """Record WHY a formula version exists. §29.

    Several drivers per version are expected: a revision may answer a
    failure and chase a requirement at once, and a single column would
    force somebody to pick one and lose the rest.
    """
    try:
        driver_id = record_driver(
            session,
            formula_version_id=formula_version_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            spec=DriverInput(**payload.model_dump()),
        )
    except FailureNotFoundError as exc:
        raise _missing(exc) from exc
    except FailureError as exc:
        raise _invalid(exc) from exc
    return {"id": str(driver_id)}

File: apps/web/lib/api/failures.ts
/**
 * Failure investigations and the approval queue, over HTTP.
 *
 * 🔴 SLICE 6 SHIPPED ITS BACKEND AND NOT ITS BROWSER, AND THIS IS THE HALF
 * THAT WAS MISSING.
 *
 * Measured 2026-08-27: `app/api/failures.py` declares **eleven write
 * endpoints** and not one of them had a client function, so a person could
 * not open an investigation, propose a hypothesis, record evidence, accept a
 * root cause, raise a corrective action or close anything. The approval queue
 * was in the same state — the only way to reach it was to construct an HTTP
 * request by hand.
 *
 * ⚠️ ONE MODULE, TWO ROUTERS, TWO PREFIXES. `app/api/failures.py` mounts
 * `router` at `/api/quality/failures` AND `approvals_router` at
 * `/api/approvals`. The first draft of this file put the approval paths under
 * the failures prefix because they live in the same source file, which is a
 * fact about the file and not about the API. Every path below was read off
 * `app/main.py`'s `include_router` calls instead.
 *
 * *A route with no caller is the same defect as a table with no writer*, and
 * this project found twenty-three of them on 2026-08-24 across four other
 * modules. These are the next eleven.
 *
 * ---------------------------------------------------------------------------
 * TWO RULES THIS FILE EXISTS TO KEEP
 * ---------------------------------------------------------------------------
 *
 * 🔴 **AN AI HYPOTHESIS IS NOT AN ACCEPTED ROOT CAUSE** (§7). Every hypothesis
 * carries `origin` (`human` | `msd`) and `status`
 * (`proposed` | `under_review` | `accepted` | `rejected`), and **only a human
 * moves anything to `accepted`** — through `POST /root-cause`, which requires
 * `failure.accept_root_cause` and a mandatory rationale. Both fields are
 * modelled here rather than dropped, because a screen that showed a
 * confidently-worded MSD hypothesis without saying where it came from is the
 * single worst thing this module can render.
 *
 * 🔴 **EVIDENCE BEARS ON A HYPOTHESIS IN THREE WAYS, NOT ONE.** The link
 * carries `supports` | `contradicts` | `inconclusive`, and `get_failure`'s own
 * docstring says why it must be shown: *"a screen showing only supporting
 * evidence would make every hypothesis look well-founded."* So `relationship`
 * is required in the schema, not optional.
 */

import { z } from "zod";

import { apiRequest, type ApiCredentials } from "./client";

// ---------------------------------------------------------------------------
// Failure investigations
// ---------------------------------------------------------------------------

/**
 * A row in the investigation queue.
 *
 * `hypothesis_count`, `has_root_cause` and `open_actions` come back on every
 * row because they are what makes the queue actionable — §11 requires counts
 * to represent items needing action, not total rows. An investigation with
 * three hypotheses and no accepted root cause is a different piece of work
 * from one with none at all, and the list is where that has to be visible.
 */
export const failureSummarySchema = z.object({
  id: z.string(),
  failure_code: z.string(),
  title: z.string(),
  severity: z.string(),
  status: z.string(),
  project_id: z.string(),
  test_id: z.string().nullable(),
  formula_version_id: z.string().nullable(),
  opened_at: z.string(),
  closed_at: z.string().nullable(),
  hypothesis_count: z.number(),
  // 🔴 A BOOLEAN, AND IT WAS A COUNT UNTIL THIS CLIENT WAS WRITTEN.
  //
  // `list_failures` returned `(SELECT count(*) …) AS has_root_cause` — a
  // column whose NAME asks a yes/no question and whose VALUE was a number.
  // Every consumer reading it loosely would have been right by accident (0 is
  // falsy, 2 is truthy) and this schema, which validates, would have rejected
  // every response. *Ask what a returned value answers, not what the column is
  // called.* Fixed in the service on 2026-08-27 with `> 0`, and pinned by
  // `test_has_root_cause_answers_the_question_its_name_asks`, so it cannot
  // quietly go back to a count.
  has_root_cause: z.boolean(),
  open_actions: z.number(),
});
export type FailureSummary = z.infer<typeof failureSummarySchema>;

const failureList = z.array(failureSummarySchema);

export function fetchFailures(
  credentials: ApiCredentials,
  signal?: AbortSignal,
): Promise<FailureSummary[]> {
  return apiRequest({ path: "/api/quality/failures", credentials, signal }, (payload) =>
    failureList.parse(payload),
  );
}

/** How a piece of evidence bears on one hypothesis. */
export const hypothesisEvidenceSchema = z.object({
  hypothesis_id: z.string(),
  evidence_id: z.string(),
  // Required, never defaulted. See the header: only rendering `supports`
  // makes every hypothesis look well-founded.
  relationship: z.enum(["supports", "contradicts", "inconclusive"]),
  note: z.string().nullable(),
  evidence_type: z.string(),
  summary: z.string(),
  origin: z.enum(["human", "msd"]),
});
export type HypothesisEvidence = z.infer<typeof hypothesisEvidenceSchema>;

export const hypothesisSchema = z.object({
  id: z.string(),
  possible_cause: z.string(),
  mechanism: z.string().nullable(),
  confidence: z.enum(["low", "medium", "high"]),
  source: z.string().nullable(),
  // §7 rests on this distinction and the screen renders it differently.
  origin: z.enum(["human", "msd"]),
  status: z.enum(["proposed", "under_review", "accepted", "rejected"]),
  accepted_by: z.string().nullable(),
  accepted_at: z.string().nullable(),
  rejection_reason: z.string().nullable(),
  proposed_by: z.string().nullable(),
  created_at: z.string(),
  evidence: z.array(hypothesisEvidenceSchema),
});
export type Hypothesis = z.infer<typeof hypothesisSchema>;

export const evidenceSchema = z.object({
  id: z.string(),
  evidence_type: z.string(),
  summary: z.string(),
  detail: z.string().nullable(),
  referenced_entity_type: z.string().nullable(),
  referenced_entity_id: z.string().nullable(),
  source_reference: z.string().nullable(),
  origin: z.enum(["human", "msd"]),
  recorded_at: z.string(),
});
export type Evidence = z.infer<typeof evidenceSchema>;

export const actionSchema = z.object({
  id: z.string(),
  action_type: z.string(),
  description: z.string(),
  status: z.string(),
  assigned_to: z.string().nullable(),
  due_date: z.string().nullable(),
  completed_at: z.string().nullable(),
  outcome: z.string().nullable(),
});
export type FailureAction = z.infer<typeof actionSchema>;

export const failureDetailSchema = z.object({
  id: z.string(),
  organization_id: z.string(),
  project_id: z.string(),
  failure_code: z.string(),
  title: z.string(),
  description: z.string().nullable(),
  severity: z.string(),
  status: z.string(),
  test_id: z.string().nullable(),
  formula_version_id: z.string().nullable(),
  batch_id: z.string().nullable(),
  opened_by: z.string().nullable(),
  opened_at: z.string(),
  closed_by: z.string().nullable(),
  closed_at: z.string().nullable(),
  closure_summary: z.string().nullable(),
  hypotheses: z.array(hypothesisSchema),
  // 🔴 THE SERVER DECIDES THIS, NOT THE SCREEN. `get_failure` computes it as
  // the hypothesis whose status is `accepted`. Re-deriving it in the browser
  // would be a second implementation of "what is the root cause", free to
  // disagree with the one the database answered — the two-literals defect
  // applied to the most consequential field in the module.
  accepted_root_cause: hypothesisSchema.nullable(),
  evidence: z.array(evidenceSchema),
  actions: z.array(actionSchema),
});
export type FailureDetail = z.infer<typeof failureDetailSchema>;

export function fetchFailure(
  credentials: ApiCredentials,
  failureId: string,
  signal?: AbortSignal,
): Promise<FailureDetail> {
  return apiRequest(
    { path: `/api/quality/failures/${failureId}`, credentials, signal },
    (payload) => failureDetailSchema.parse(payload),
  );
}

// ---------------------------------------------------------------------------
// The writes. Eleven endpoints, none of which had a caller.
// ---------------------------------------------------------------------------

export interface HypothesisRequest {
  readonly possible_cause: string;
  readonly mechanism?: string;
  readonly confidence: "low" | "medium" | "high";
  readonly source?: string;
}

/**
 * Propose a hypothesis.
 *
 * ⚠️ `origin` IS NOT IN THIS REQUEST, AND THAT IS THE POINT. The server
 * defaults it to `human` and the field exists so an MSD-proposed hypothesis
 * must SAY it is one. A browser form that could send `origin: "msd"` would let
 * a person file a machine's opinion under the machine's name, or — worse — let
 * a future MSD path file its own under a human's. The label can only be wrong
 * by a caller asserting it, so this caller cannot assert it.
 */
export function proposeHypothesis(
  credentials: ApiCredentials,
  failureId: string,
  request: HypothesisRequest,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/quality/failures/${failureId}/hypotheses`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => payload,
  );
}

export interface EvidenceRequest {
  readonly evidence_type: string;
  readonly summary: string;
  readonly detail?: string;
  readonly source_reference?: string;
}

/** Record evidence against the investigation. `origin` is the server's, as above. */
export function recordEvidence(
  credentials: ApiCredentials,
  failureId: string,
  request: EvidenceRequest,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/quality/failures/${failureId}/evidence`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => payload,
  );
}

export interface EvidenceLinkRequest {
  readonly evidence_id: string;
  readonly relationship: "supports" | "contradicts" | "inconclusive";
  readonly note?: string;
}

/** Link a piece of evidence to a hypothesis, stating HOW it bears on it. */
export function linkEvidence(
  credentials: ApiCredentials,
  failureId: string,
  hypothesisId: string,
  request: EvidenceLinkRequest,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/quality/failures/${failureId}/hypotheses/${hypothesisId}/evidence`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => payload,
  );
}

/**
 * Accept a hypothesis as the root cause. **A human act, and only a human act.**
 *
 * `rationale` is mandatory server-side and mandatory in this signature:
 * accepting a root cause is a technical decision and an unexplained one cannot
 * be reviewed later. Requires `failure.accept_root_cause`, held by the Lead and
 * the Director — not by the investigator who proposed it.
 */
export function acceptRootCause(
  credentials: ApiCredentials,
  failureId: string,
  request: { readonly hypothesis_id: string; readonly rationale: string },
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/quality/failures/${failureId}/root-cause`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => payload,
  );
}

/** Reject a hypothesis, with a reason that stays on the record. */
export function rejectHypothesis(
  credentials: ApiCredentials,
  failureId: string,
  hypothesisId: string,
  reason: string,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/quality/failures/${failureId}/hypotheses/${hypothesisId}/rejection`,
      method: "POST",
      body: { reason },
      credentials,
    },
    (payload) => payload,
  );
}

export interface ActionRequest {
  readonly action_type: string;
  readonly description: string;
  readonly due_date?: string;
}

/** Raise a corrective action. */
export function raiseAction(
  credentials: ApiCredentials,
  failureId: string,
  request: ActionRequest,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/quality/failures/${failureId}/actions`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => payload,
  );
}

/** Close the investigation. Requires `failure.close`. */
export function closeFailure(
  credentials: ApiCredentials,
  failureId: string,
  summary: string,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/quality/failures/${failureId}/closure`,
      method: "POST",
      body: { summary },
      credentials,
    },
    (payload) => payload,
  );
}

// ---------------------------------------------------------------------------
// The approval queue
// ---------------------------------------------------------------------------

/**
 * One step this caller could decide **right now**.
 *
 * "Right now" is load-bearing and the server owns it: `pending_steps_for`
 * excludes steps whose turn has not come, because §11 requires a queue to
 * represent items needing action BY THE HOLDER. The screen must not filter
 * this further and must not widen it — both would be a second implementation
 * of whose turn it is.
 */
export const approvalQueueItemSchema = z.object({
  step_id: z.string(),
  step_number: z.number(),
  step_label: z.string(),
  permission_required: z.string().nullable(),
  route_id: z.string(),
  entity_type: z.string(),
  entity_id: z.string(),
  template_code: z.string(),
  project_id: z.string(),
  opened_at: z.string(),
});
export type ApprovalQueueItem = z.infer<typeof approvalQueueItemSchema>;

const approvalQueue = z.array(approvalQueueItemSchema);

export function fetchApprovalQueue(
  credentials: ApiCredentials,
  signal?: AbortSignal,
): Promise<ApprovalQueueItem[]> {
  return apiRequest({ path: "/api/approvals/queue", credentials, signal }, (payload) =>
    approvalQueue.parse(payload),
  );
}

export const routeStepSchema = z.object({
  id: z.string(),
  step_number: z.number(),
  parallel_group: z.number(),
  permission_required: z.string().nullable(),
  step_label: z.string(),
  is_mandatory: z.boolean(),
  must_differ_from_group: z.boolean(),
  decision: z.string().nullable(),
  condition_text: z.string().nullable(),
  rationale: z.string().nullable(),
  decided_by: z.string().nullable(),
  decided_at: z.string().nullable(),
});
export type RouteStep = z.infer<typeof routeStepSchema>;

export const approvalRouteSchema = z.object({
  id: z.string(),
  entity_type: z.string(),
  entity_id: z.string(),
  template_code: z.string(),
  status: z.string(),
  project_id: z.string(),
  opened_at: z.string(),
  closed_at: z.string().nullable(),
  steps: z.array(routeStepSchema),
  // 🔴 COMPUTED BY THE ENGINE, NOT BY THE SCREEN. `get_route`'s docstring is
  // explicit that a stored "current step" would go stale the moment a parallel
  // step was decided out of order, so it is derived per read. A browser that
  // worked out whose turn it is would be the stale copy the server refused to
  // keep.
  current_group: z.number().nullable(),
  next_steps: z.array(routeStepSchema),
  awaiting: z.array(z.string()),
});
export type ApprovalRoute = z.infer<typeof approvalRouteSchema>;

export function fetchApprovalRoute(
  credentials: ApiCredentials,
  routeId: string,
  signal?: AbortSignal,
): Promise<ApprovalRoute> {
  return apiRequest(
    { path: `/api/approvals/${routeId}`, credentials, signal },
    (payload) => approvalRouteSchema.parse(payload),
  );
}

export interface StepDecisionRequest {
  readonly decision:
    | "approve"
    | "approve_with_condition"
    | "return_for_correction"
    | "request_retest"
    | "reject"
    | "escalate"
    | "request_additional_test";
  readonly condition_text?: string;
  readonly rationale?: string;
}

/**
 * Record a decision on one rung.
 *
 * ⚠️ `require_permission("test.view")` ON THE ROUTE IS ONLY A FLOOR, and the
 * engine re-checks the STEP's own `permission_required` plus the
 * segregation-of-duties rule. So a 403 here means one of two different things
 * — you lack the rung's permission, or you hold it and are barred by your own
 * earlier involvement — and the server says which. The message is surfaced
 * verbatim rather than translated.
 */
export function decideStep(
  credentials: ApiCredentials,
  routeId: string,
  stepId: string,
  request: StepDecisionRequest,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/approvals/${routeId}/steps/${stepId}/decision`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => payload,
  );
}
